building-and-packaging/material/packaging_hpc_notes.md

```diff
@@ -0,0 +1,196 @@
+
+      # FIXME: Add a proper url for your package's homepage here.
+      homepage = "https://www.example.com"
+      url      = "https://github.com/Simulation-Software-Engineering/HelloWorld/archive/refs/tags/v0.3.0.tar.gz"
+
+      # FIXME: Add a list of GitHub accounts to
+      # notify when the package is updated.
+      # maintainers = ['github_user1', 'github_user2']
+
+      version('0.3.0', sha256='5e95bf1904ebfbd5c28b064bff098a3bb654bf7c407f2031295e3588d6d9e8fa')
+      version('0.2.0', sha256='a77499bbfd0b8f4d7070b06c491e062fa608fdd7e939d6c37796bdafdbbaa35a')
+      version('0.1.0', sha256='2e99e2ff2b955b9d17645367a0a7eeb485162d9336cdbf0034b9d95d464f3157')
+
+      # FIXME: Add dependencies if required.
+      # depends_on('foo')
+
+      def cmake_args(self):
+          # FIXME: Add arguments other than
+          # FIXME: CMAKE_INSTALL_PREFIX and CMAKE_BUILD_TYPE
+          # FIXME: If not needed delete this function
+          args = []
+          return args
+  ```
+
+  At least the `FIXME` statements should be fixed before relasing the package. The Spack package recipe is written in Python so we can also do common Python magic here.
+
+- Spack has deduced a good number of information already.
+    - We work with CMake
+    - There are 3 releases of the software
+    - The URL to download packages
+    - The name of the Package we work with `class Helloworld`. This is also the name of the software now within Spack.
+- We want to fix/extend the package with some standard information
+    - Package description
+    - Set URL to SSE homepage
+    - Add our GitHub username as maintainer
+    - Remove the `cmake_args` part as we only have a standard CMake arguments. Here could give extra/special arguments specific to the software package.
+- Concretize our package
+
+  ```bash
+  $ spack spec helloworld
+  Input spec
+  --------------------------------
+  helloworld
+
+  Concretized
+  --------------------------------
+  helloworld@0.3.0%gcc@9.3.0~ipo build_type=RelWithDebInfo arch=linux-ubuntu20.04-skylake
+    ^cmake@3.16.3%gcc@9.3.0~doc+ncurses+openssl+ownlibs~qt build_type=Release patches=1c540040c7e203dd8e27aa20345ecb07fe06570d56410a24a266ae570b1c4c39,bf695e3febb222da2ed94b3beea600650e4318975da90e4a71d6f31a6d5d8c3d arch=linux-ubuntu20.04-skylake
+  ```
+
+  We see that `cmake` is an implicit dependency as we need it for building our package.
+
+- Install package
+
+  ```bash
+  spack install helloworld
+  ```
+
+- Load installed package, run code and unload
+
+  ```bash
+  spack load helloworld
+  helloworld
+  spack unload helloworld
+  ```
+
+- Install different version of code
+
+  ```bash
+  spack install helloworld@0.2.0
+  ```
+
+  This will concretize (internally, i.e. no output on terminal) and then build the software.
+
+- Add `main` branch and thus GitHub repository
+
+  ```diff
+  + git      = "https://github.com/Simulation-Software-Engineering/HelloWorld.git"
+  +
+  + version('main', branch='main')
+  ```
```

Comment:
If you want the most recent version of a package, one would want to add the Git repository. This could be the case if some other package has our package as dependency and needs a new fix. Another use case is code development in Spack where on wants to work with Git branches.

By default, the Spack package will only look for releases that one can download. At least I have used it some times and found this feature useful.

Reply:
OK. I would suggest to either add some comments about this or to use the feature?

Reply:
I will remove it for now... the slides are full enough.